crates/ty_python_semantic/resources/mdtest/ide_support/all_members.md

```diff
@@ -0,0 +1,239 @@
+
+class C(Intermediate):
+    class_attr: int = 3
+
+    def f_c(self):
+        self.instance_attr = "C"
+
+    @property
+    def property_attr(self) -> int:
+        return 1
+
+members_of_instance = all_members(C())
+
+static_assert("base_class_attr" in members_of_instance)
+static_assert("intermediate_attr" in members_of_instance)
+static_assert("class_attr" in members_of_instance)
+
+static_assert("base_instance_attr" in members_of_instance)
+static_assert("intermediate_instance_attr" in members_of_instance)
+static_assert("instance_attr" in members_of_instance)
+
+static_assert("f_base" in members_of_instance)
+static_assert("f_intermediate" in members_of_instance)
+static_assert("f_c" in members_of_instance)
+
+static_assert("property_attr" in members_of_instance)
+
+static_assert("non_existent" not in members_of_instance)
+```
+
+### Class objects
+
+Class-level attributes can also be accessed through the class itself:
+
+```py
+from ty_extensions import all_members, static_assert
+
+class Base:
+    base_attr: int = 1
+
+class C(Base):
+    class_attr: str = "c"
+
+    def f(self):
+        self.instance_attr = True
+
+members_of_class = all_members(C)
+
+static_assert("class_attr" in members_of_class)
+static_assert("base_attr" in members_of_class)
+
+static_assert("non_existent" not in members_of_class)
+```
+
+But instance attributes can not be accessed this way:
+
+```py
+static_assert("instance_attr" not in members_of_class)
+```
+
+When a class has a metaclass, members of that metaclass (and bases of that metaclass) are also
+accessible:
+
+```py
+class MetaBase(type):
+    meta_base_attr = 1
+
+class Meta(MetaBase):
+    meta_attr = 2
+
+class D(Base, metaclass=Meta):
+    class_attr = 3
+
+static_assert("meta_base_attr" in all_members(D))
+static_assert("meta_attr" in all_members(D))
+static_assert("base_attr" in all_members(D))
+static_assert("class_attr" in all_members(D))
+```
+
+### Generic classes
+
+```py
+from ty_extensions import all_members, static_assert
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+class C(Generic[T]):
+    base_attr: T
+
+static_assert("base_attr" in all_members(C[int]))
+static_assert("base_attr" in all_members(C[int]()))
+```
+
+### Other instance-like types
+
+```py
+from ty_extensions import all_members, static_assert
+from typing_extensions import LiteralString
+
+static_assert("__xor__" in all_members(True))
+static_assert("bit_length" in all_members(1))
+static_assert("startswith" in all_members("a"))
+static_assert("__buffer__" in all_members(b"a"))
+
+def _(literal_string: LiteralString):
+    static_assert("startswith" in all_members(literal_string))
+
+static_assert("count" in all_members(("some", "tuple", 1, 2)))
+
+static_assert("__doc__" in all_members(len))
+static_assert("__doc__" in all_members("a".startswith))
+```
+
+### Unions
+
+For unions, `all_members` will only return members that are available on all elements of the union.
```

Comment:
This might be an interesting correctness-vs-UX tradeoff. When given a `obj: str | None` object, for example, it's technically correct to only show attributes that are available on `str` *and* `None` (so basically nothing, except for a few special methods on `object` that nobody wants to call directly).

But the more likely scenario is probably that the developer actually wants to access an attribute on `str`, and just forgot to eliminate the `None` possibility. So it might be more useful to show attributes that ar available on `str` *or* `None` here instead? Not sure.

Reply:
Well I would change that docstring obviously, *if* we would decide to deliberate return members that are not available :smile: This function is only here to support that precise use case (providing completion results for the LPS). Making it available via `ty_extensions.all_members` was just a convenient way for me to test it. I'm not sure if this should really be exposed. Or if we should maybe move it to `ty_extensions.internal` or similar.

Reply:
I'm inclined to stick to what you have here for now (preferring correctness), and then loosen it up as use cases arise. Maybe, e.g., it makes sense to only loosen it in certain scenarios.

Reply:
For reference, Pyright / Pylance does provide all members of the union type. In the following example, `decode` is only present on `bytes` while other methods are present on both:

<img width="521" alt="Screenshot 2025-05-30 at 10 06 56" src="https://github.com/user-attachments/assets/386be55e-90c5-4e44-aacb-e55bce7e30a3" />

Reply:
I suspect that this will require some iteration and/or even configurations. E.g. r-a has an option that allows you to configure if r-a should hide members to which you don't have access to due to visibility constraint. Hiding them is probably the right default when working with third-party code, but for first-party code, I often want to see all members, and I'll then change the visibility of the member.

So I think we should move forward here as we can change this easily in the future

Reply:
Yeah I think this should be easy to change.

I can imagine wanting configuration for respecting `__all__` too.